jonason/Tokenizer.cpp

```diff
@@ -0,0 +1,103 @@
+//
+//  Tokenizer.cpp
+//  jonason
+//
+//  Created by Gabriel Lumbi on 2023-03-02.
+//
+
+#include "Tokenizer.hpp"
+
+#include <sstream>
+
+namespace jonason {
+
+Token::Token(Token&& other) {
+    tag = std::move(other.tag);
+    value = std::move(other.value);
+    other.value = nullptr;
+}
+
+Token& Token::operator=(Token&& other) {
+    tag = std::move(other.tag);
+    value = std::move(other.value);
+    other.value = nullptr;
+    return *this;
+}
+
+Token::~Token() {
+    delete[] value;
+}
+
+constexpr bool is_ws(char character) {
+    return character == 0x20 || character == 0x0a || character == 0x0d || character == 0x09;
+}
+
+static void read_literal(std::istream& istream, char*& buffer, int& buffer_size, std::predicate<char> auto is_delimiter, std::vector<Token>& out);
+
+void tokenize(const std::string& string, std::vector<Token>& out)
+{
+    std::istringstream istream(string);
+    tokenize(istream, out);
+}
+
+const int MAX_LITERAL_SIZE = 1 << 16; // about 65KB
+
+void tokenize(std::istream& istream, std::vector<Token>& out) {
+    char char_buffer;
+    char* value_buffer = new char[MAX_LITERAL_SIZE];
+    int value_buffer_count = 0;
+
+    istream >> std::ws;
```

Comment:
That's new to me. I hope it uses the same definition of whitespace as your `is_ws`.